### PR TITLE
Fix type hints of tuner/batch_size_scaling.py

### LINKED ISSUE #13445: Improve typing coverage (4/n)
## 🚀 Typing coverage

Let's improve typing coverage of PyTorch Lightning together!

I'm creating a new issue in order to increase visibility. There are three older issues (#7037, #5023, #4698) which became stale over time.

### Plan

Currently, there are 55 files which are excluded from mypy checks so that our CI does not fail. These files vastly differ in difficulty in order to make the typing complete. For this reason, we are introducing difficulty estimate for each file so that community members can choose to work on the files appropriate to their skill level.

Please, comment on this issue in order to reserve a particular file to work on. Once you do so, I will edit this top comment to avoid collisions. Once you think your work is finished, please open a PR referencing this issue which:

 * removes the corresponding line from `pyproject.toml`
 * and passes `mypy` checks with the corresponding line removed. You can test it locally by running `mypy` from root directory

If you are struggling with pushing it over the finish line, open the PR anyway and someone from our team will help you to get it there. 🚀

Please note, that it can happen that you may need to edit more than just one file. This is fine, but please keep in mind, that the goal of your PR will be to make the check passing for the chosen file. Also, please note that the difficulty is just an educated guess.

For those of you who are not familiar with the process of contributing a PR, we have prepared a [simple guide](https://devblog.pytorchlightning.ai/quick-contribution-guide-86d977171b3a) that will walk you through the necessary steps. You can do it! :rocket: :muscle: 

#### Difficulty 3 of 3

- [x] `pytorch_lightning/callbacks/progress/rich_progress.py` @donlapark #14963 

### Completed

<details>

#### Difficulty 1 of 3

- [x] `pytorch_lightning/core/decorators.py` #14044
- [x] `pytorch_lightning/profilers/advanced.py` @nninept #13792 
~- [ ] `pytorch_lightning/profilers/base.py` @LeeChanHyuk #13879~
- [x] `pytorch_lightning/loggers/base.py` @JustinGoheen #13494 
- [x] `pytorch_lightning/__setup__.py` @CyprienRicque #13472 
~- [ ] `pytorch_lightning/distributed/dist.py` @puhuk #13492~
- [x] `pytorch_lightning/strategies/single_device.py` @CyprienRicque #13532 
- [x] `pytorch_lightning/trainer/optimizers.py` @gautierdag #13470
- [x] `pytorch_lightning/utilities/distributed.py` @krishnakalyan3 #13678
- [x] `pytorch_lightning/callbacks/finetuning.py` @ar90n #13516 
- [x] `pytorch_lightning/loggers/mlflow.py` @JustinGoheen ~~#13690~~ #13691
- [x] `pytorch_lightning/tuner/tuning.py` @donlapark ~~#13616~~ #13631 
- [x] `pytorch_lightning/strategies/single_tpu.py` @CyprienRicque #13534 
- [x] `pytorch_lightning/strategies/ddp2.py` @CyprienRicque #13535 
- [x] `pytorch_lightning/strategies/parallel.py` @CyprienRicque #13556 
- [x] `pytorch_lightning/loggers/csv_logs.py` @JustinGoheen #13538 
- [x] `pytorch_lightning/tuner/lr_finder.py` @donlapark #13513 #13652 
- [x] `pytorch_lightning/strategies/dp.py` @CyprienRicque #13564 
- [x] `pytorch_lightning/profilers/simple.py` @krishnakalyan3 #14103
- [x] `pytorch_lightning/strategies/sharded_spawn.py` @krishnakalyan3 #14102
- [x] `pytorch_lightning/demos/mnist_datamodule.py` @alro923 #13929 
- [x] `pytorch_lightning/demos/boring_classes.py` @krishnakalyan3 #14201
- [x] `pytorch_lightning/tuner/batch_size_scaling.py` @ar90n #13518 

#### Difficulty 2 of 3

- [x] `pytorch_lightning/loops/epoch/training_epoch_loop.py` @himkt #13555 
- [x] `pytorch_lightning/core/mixins/device_dtype_mixin.py` @krishnakalyan3 #13704 
- [x] `pytorch_lightning/loggers/comet.py` @JustinGoheen #13689 
- [x] `pytorch_lightning/loggers/tensorboard.py` @JustinGoheen #13688 
- [x] `pytorch_lightning/strategies/horovod.py` @CyprienRicque #13570 
- [x] `pytorch_lightning/callbacks/model_checkpoint.py` @BongYang #13617
- [x] `pytorch_lightning/strategies/fully_sharded.py` @BongYang #13941 
- [x] `pytorch_lightning/loggers/neptune.py` @JustinGoheen #13692 
- [x] `pytorch_lightning/utilities/meta.py` @nninept #13763 #13868
- [x] `pytorch_lightning/strategies/tpu_spawn.py` @BongYang #13813  
- [x] `pytorch_lightning/loggers/logger.py` @JustinGoheen #13541 
- [x] `pytorch_lightning/loggers/wandb.py` @gautierdag #13483
- [x] `pytorch_lightning/callbacks/stochastic_weight_avg.py` @donlapark #13685 #13860
- [x] `pytorch_lightning/strategies/strategy.py` @CyprienRicque #13519 
- [x] `pytorch_lightning/strategies/deepspeed.py` @donlapark #13832 
- [x] `pytorch_lightning/strategies/ddp_spawn.py` @donlapark #13865 
- [x] `pytorch_lightning/strategies/ipu.py` @HalestormAI #13786
- [x] `pytorch_lightning/trainer/connectors/callback_connector.py` @krishnakalyan3 #13750
- [x] `pytorch_lightning/strategies/ddp.py` @lijm1358 #13885 
- [x] `pytorch_lightning/core/saving.py` @JustinGoheen #13932 
- [x] `pytorch_lightning/callbacks/quantization.py` @krishnakalyan3 #13782 
- [x] `pytorch_lightning/strategies/sharded.py` @lijm1358 #14184 
- [x] `pytorch_lightning/core/datamodule.py` @JustinGoheen #13693

#### Difficulty 3 of 3
~- [ ] `pytorch_lightning/trainer/callback_hook.py` @JustinGoheen #13807 ~
- [x] `pytorch_lightning/core/module.py` @JustinGoheen #13603 
- [x] `pytorch_lightning/trainer/connectors/data_connector.py` @JustinGoheen #13806
- [x] `pytorch_lightning/utilities/auto_restart.py` @donlapark #13904 
- [x] `pytorch_lightning/trainer/supporters.py` @donlapark #14633 
- [x] `pytorch_lightning/profilers/pytorch.py` @krishnakalyan3 #14405 
- [x] `pytorch_lightning/utilities/data.py` @nandwalritik #13901 
- [x] `pytorch_lightning/trainer/trainer.py` ~@JustinGoheen #13810~ @BongYang #14204 

</details>

cc @borda @justusschock @awaelchli @rohitgr7 @Borda @tchaton @aniketmaurya @kingjuno @alat-rights @carmocca @akihironitta @stancld as you were all involved in previous issues

### PR DESCRIPTION
**Pull request by ar90n**
## What does this PR do?
If wrong parameters are given as arguments, `_adjust_batch_size` cannot adjust batch size correctly. So this PR add some codes to raise ValueErrors in that cases.

Fixes part of #13445

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/pytorch_lightning/utilities/data.py
Comment (Borda) [inline src/pytorch_lightning/utilities/data.py:129]: shall this be at least an Iterable type?

File: src/pytorch_lightning/tuner/batch_size_scaling.py
Comment (awaelchli) [inline src/pytorch_lightning/tuner/batch_size_scaling.py:239]:
```suggestion
            raise ValueError(f"Batch size attribute in LightningModule must be an integer, got: {batch_size!r}")
```

File: src/pytorch_lightning/utilities/data.py
Comment (ar90n) [reply src/pytorch_lightning/utilities/data.py]: Thanks for your review and comment. After I tried some experiments, I found that this modification is unnecessary for this PR. So I reverted this line.

**Review by otaj — APPROVED** [review]

File: src/pytorch_lightning/tuner/batch_size_scaling.py
Comment (justusschock) [inline src/pytorch_lightning/tuner/batch_size_scaling.py:240]:
```suggestion
    new_size: int = value if value is not None else int(batch_size * factor)
```
that doesn't work?

**Review by justusschock — APPROVED** [review]

File: src/pytorch_lightning/tuner/batch_size_scaling.py
Comment (rohitgr7) [inline src/pytorch_lightning/tuner/batch_size_scaling.py:256]: this is not the right way. there should be a dataloader to check the initial value. Fixed it #11089

**Review by rohitgr7 — COMMENTED** [review]
can we close this PR in favor of #11089?

File: src/pytorch_lightning/tuner/batch_size_scaling.py
Comment (ar90n) [reply src/pytorch_lightning/tuner/batch_size_scaling.py:240]: `batch_size` may be None. So mypy says error about `*` not supporting with `None` and `float` .

File: src/pytorch_lightning/tuner/batch_size_scaling.py
Comment (ar90n) [reply src/pytorch_lightning/tuner/batch_size_scaling.py:256]: Thanks for your suggestion. I got it.

**Review by carmocca — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

File: src/pytorch_lightning/tuner/batch_size_scaling.py
Comment (otaj) [inline src/pytorch_lightning/tuner/batch_size_scaling.py:242]: Hi @ar90n , it seems that tests started failing on this check. Can you please take a look?

**Review by otaj — APPROVED** [review]

File: src/pytorch_lightning/tuner/batch_size_scaling.py
Comment (ar90n) [reply src/pytorch_lightning/tuner/batch_size_scaling.py:242]: Hi @otaj. In my understanding, these errors will be fixed in  #11089. So I think that this PR will be merged after #11089 is merged or will be closed without merging.

I'm so sorry if my understanding is wrong. If I should fix these errors. I'm going to try. But if so, this PR may contain a lot of modifications that are not for type hints.

Comment (otaj) [reply src/pytorch_lightning/tuner/batch_size_scaling.py:242]: Hi @ar90n, you are absolutely correct, sorry about the noise. I should keep myself more up to date with other PRs :D

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
@otaj mind checking out... :otter:

**Comment by otaj** [comment]
> @otaj mind checking out... :otter:

I think that this is currently blocked by #11089 - and since that one is not merged yet, this cannot be merged as well. Is that correct, @ar90n?

**Comment by rohitgr7** [comment]
> > @otaj mind checking out... 🦦
> 
> I think that this is currently blocked by #11089 - and since that one is not merged yet, this cannot be merged as well. Is that correct, @ar90n?

yes!

**Comment by ar90n** [comment]
@otaj 
Yes, I agree. After https://github.com/Lightning-AI/lightning/pull/11089 will be merged, I will fix some conflicts if needed.

**Comment by otaj** [comment]
Hi, @ar90n, #11089 just got merged :rocket:

**Review by otaj — APPROVED** [review]

File: src/pytorch_lightning/callbacks/batch_size_finder.py
Comment (carmocca) [inline src/pytorch_lightning/callbacks/batch_size_finder.py:34]: Why is this annotated in the class definition instead of in the `__init__`?

Comment (ar90n) [reply src/pytorch_lightning/callbacks/batch_size_finder.py:34]: @carmocca 
This is my mistake. And my modification wasn't in time.
In the head of my fork branch, this annotation is moved into __init__.
May I create a new PR to move this annotation into __init__ ?

https://github.com/ar90n/pytorch-lightning/commit/7076c1bfc942710d48793a866e3622d8c43f5632

Comment (carmocca) [reply src/pytorch_lightning/callbacks/batch_size_finder.py:34]: Yes, please. Thank you :purple_heart:

**Pull request merged**
